### PR TITLE
Update Azure SDK

### PR DESCRIPTION
**Pull request by donatello**
## Description

Update Azure SDK in preparation for copy-object-with-metadata fix (#4924).

This change refactors the approach of translating S3 headers to Azure headers as the Azure SDK requires separate BlobMetadata and BlobProperties structures (where the former will include only user defined custom metadata, and the latter will have common properties such as Content-Encoding, etc).

## Motivation and Context

Need to update library to latest version to enable the fix for #4924.

## How Has This Been Tested?

Minio-Java functional tests were run.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: cmd/gateway-azure.go [hunk 1]
```diff
@@ -41,52 +41,112 @@ const globalAzureAPIVersion = "2016-05-31"
...
 	gatewayHeaders := map[string]string{
 		"X-Amz-Meta-X-Amz-Key":     "X-Amz-Meta-x_minio_key",
 		"X-Amz-Meta-X-Amz-Matdesc": "X-Amz-Meta-x_minio_matdesc",
 		"X-Amz-Meta-X-Amz-Iv":      "X-Amz-Meta-x_minio_iv",
 	}
 
-	newHeaders = make(map[string]string)
-	for k, v := range headers {
+	blobMeta := make(map[string]string)
+	var props storage.BlobProperties
```
Comment (balamurugana) [inline cmd/gateway-azure.go]: You could make `blobMeta` and `props` in the definition like `(metadata storage.BlobMetadata, properties storage.BlobProperties)`

File: cmd/gateway-azure.go [hunk 2]
```diff
@@ -41,52 +41,112 @@ const globalAzureAPIVersion = "2016-05-31"
...
+// automatically prefixed with `X-Ms-Meta-` by the Azure SDK. S3
+// user-metadata is translated to Azure metadata by removing the
+// `X-Amz-Meta-` prefix.
+//
+// BlobProperties contains commonly set metadata for objects such as
+// Content-Encoding, etc. Such metadata that is accepted by S3 is
+// copied into BlobProperties.
+//
+// Header names are canonicalized as in http.Header.
+func s3MetaToAzure(metadata map[string]string) (storage.BlobMetadata, storage.BlobProperties) {
```
Comment (balamurugana) [inline cmd/gateway-azure.go]: 1. you could rename this function as `s3MetatoAzureProperties()`
2. `metadata` can be `s3Metadata`

File: cmd/gateway-azure.go [hunk 3]
```diff
@@ -41,52 +41,112 @@ const globalAzureAPIVersion = "2016-05-31"
...
-		"X-Amz-Meta-x_minio_matdesc": "X-Amz-Meta-X-Amz-Matdesc",
-		"X-Amz-Meta-x_minio_iv":      "X-Amz-Meta-X-Amz-Iv",
+		"x-amz-meta-x_minio_key":     "X-Amz-Meta-X-Amz-Key",
+		"x-amz-meta-x_minio_matdesc": "X-Amz-Meta-X-Amz-Matdesc",
+		"x-amz-meta-x_minio_iv":      "X-Amz-Meta-X-Amz-Iv",
 	}
 
-	newMeta = make(map[string]string)
-
+	result := make(map[string]string)
```
Comment (balamurugana) [inline cmd/gateway-azure.go]: `result` can be `s3Metadata`

File: cmd/gateway-azure.go [hunk 4]
```diff
@@ -41,52 +41,112 @@ const globalAzureAPIVersion = "2016-05-31"
...
 	for k, v := range meta {
-		k = "X-Amz-Meta-" + k
+		// k's `x-ms-meta-` prefix is already stripped by
+		// Azure SDK, so we add the AMZ prefix.
+		k = "x-amz-meta-" + strings.ToLower(k)
```
Comment (balamurugana) [inline cmd/gateway-azure.go]: Have you seen any problem/misbehavior with current code of handling `k`?   Basically I am not favor of `strings.ToLower()`

File: cmd/gateway-azure.go [hunk 5]
```diff
@@ -427,49 +489,38 @@ func (a *azureObjects) GetObject(bucket, object string, startOffset int64, lengt
...
-	err = a.client.CreateBlockBlobFromReader(bucket, object, uint64(data.Size()), data, s3ToAzureHeaders(metadata))
+	blob := a.client.GetContainerReference(bucket).GetBlobReference(object)
+	blob.Metadata, blob.Properties = s3MetaToAzure(metadata)
+	err = blob.CreateBlockBlobFromReader(data, nil)
 	if err != nil {
 		return objInfo, azureToObjectError(traceError(err), bucket, object)
 	}
 	if err = data.Verify(); err != nil {
-		a.client.DeleteBlob(bucket, object, nil)
+		_ = blob.Delete(nil)
```
Comment (balamurugana) [inline cmd/gateway-azure.go]: We would need to log the failure here

File: cmd/gateway-azure.go [hunk 6]
```diff
@@ -579,14 +635,16 @@ func (a *azureObjects) PutObjectPart(bucket, object, uploadID string, partID int
...
+		err = blob.PutBlockWithLength(id, uint64(subPartSize), io.LimitReader(data, subPartSize), nil)
 		if err != nil {
 			return info, azureToObjectError(traceError(err), bucket, object)
 		}
 		subPartNumber++
 	}
 	if err = data.Verify(); err != nil {
-		a.client.DeleteBlob(bucket, object, nil)
+		blob := a.client.GetContainerReference(bucket).GetBlobReference(object)
+		_ = blob.Delete(nil)
```
Comment (balamurugana) [inline cmd/gateway-azure.go]: As the error from `blob.Delete()` is not returned, we would need to log the error

File: cmd/gateway-azure.go [hunk 7]
```diff
@@ -705,31 +767,25 @@ func (a *azureObjects) CompleteMultipartUpload(bucket, object, uploadID string,
...
 		if err != nil {
 			return objInfo, azureToObjectError(traceError(err), bucket, object)
 		}
 	}
 	return a.GetObjectInfo(bucket, object)
 }
 
-// Copied from github.com/Azure/azure-sdk-for-go/storage/blob.go
+// Copied from github.com/Azure/azure-sdk-for-go/storage/container.go
 func azureListBlobsGetParameters(p storage.ListBlobsParameters) url.Values {
```
Comment (balamurugana) [inline cmd/gateway-azure.go]: As this function is used in `gateway-azure-anonymous.go`, you could move it there.

File: cmd/gateway-azure.go [hunk 8]
```diff
@@ -41,52 +41,112 @@ const globalAzureAPIVersion = "2016-05-31"
...
 }
 
-// Prefix user metadata with "X-Amz-Meta-".
-// client.GetBlobMetadata() already strips "X-Ms-Meta-"
-func azureToS3Metadata(meta map[string]string) (newMeta map[string]string) {
+// azureToS3Meta converts Azure metadata/properties to S3 metadata. It
+// is the reverse of s3MetaToAzure. Azure's `.GetMetadata()`
+// lower-cases all header keys, so this is taken into account by this
+// function.
+func azureToS3Meta(meta storage.BlobMetadata, props storage.BlobProperties) map[string]string {
```
Comment (balamurugana) [inline cmd/gateway-azure.go]: you could rename this function as `azurePropertiesToS3Meta()`

**Review by balamurugana — CHANGES_REQUESTED** [review]

**Review by harshavardhana — CHANGES_REQUESTED** [review]
Keep :-1: until @balamurugana comments are addressed.

File: cmd/gateway-azure.go (hunk 7 above)
Comment (donatello) [reply cmd/gateway-azure.go]: Done.

File: cmd/gateway-azure.go (hunk 6 above)
Comment (donatello) [reply cmd/gateway-azure.go]: Done.

File: cmd/gateway-azure.go (hunk 5 above)
Comment (donatello) [reply cmd/gateway-azure.go]: Done.

File: cmd/gateway-azure.go (hunk 4 above)
Comment (donatello) [reply cmd/gateway-azure.go]: No, that was a mistake - removed it. The azure sdk already returns the keys lower-cased.

File: cmd/gateway-azure.go (hunk 3 above)
Comment (donatello) [reply cmd/gateway-azure.go]: Done.

File: cmd/gateway-azure.go (hunk 8 above)
Comment (donatello) [reply cmd/gateway-azure.go]: Done.

File: cmd/gateway-azure.go (hunk 1 above)
Comment (donatello) [reply cmd/gateway-azure.go]: Done.

File: cmd/gateway-azure.go (hunk 2 above)
Comment (donatello) [reply cmd/gateway-azure.go]: Done.

**Comment by donatello** [comment]
@balamurugana Thanks for the review. Comments addressed and I think it looks better now.

@harshavardhana Can you please put back @balamurugana as reviewer on this? He has already started reviewing.

**Review by harshavardhana — APPROVED** [review]

**Review by balamurugana — APPROVED** [review]
LGTM

**Comment by balamurugana** [comment]
Tested

**Review by krishnasrinivas — APPROVED** [review]

**Comment by deekoder** [comment]
since this is gateway, ignoring code coverage.

**Pull request merged**